sympy/core/dispatchers_add.py

```diff
@@ -0,0 +1,103 @@
+from sympy.multipledispatch import dispatch
+from sympy.calculus.util import AccumBounds
+from sympy import Basic, Expr, Number, Order, Add, Mul, Pow, S
+from sympy.matrices.expressions import MatrixExpr
+from sympy.core.numbers import ComplexInfinity
+
+
+class AddBuilder(object):
+    def __init__(self):
+        self.terms = {}
+        self.coeff = S.Zero
+        self.order_factors = []
+
+
+def append_arg_c_s(data, c, s):
+    # now we have:
+    # o = c*s, where
+    #
+    # c is a Number
+    # s is an expression with number factor extracted
+    # let's collect terms with the same s, so e.g.
+    # 2*x**2 + 3*x**2  ->  5*x**2
+    if s in data.terms:
+        data.terms[s] += c
+        if data.terms[s] is S.NaN:
+            # we know for sure the result will be nan
+            return S.NaN
+    else:
+        data.terms[s] = c
+
+
+@dispatch(AddBuilder, Expr)
+def append_arg_Add(data, o):
+    # everything else
+    c = S.One
+    s = o
+    return append_arg_c_s(data, c, s)
+
+@dispatch(AddBuilder, Order)
+def append_arg_Add(data, o):
+    for o1 in data.order_factors:
+        if o1.contains(o):
+            o = None
+            break
+    if o is None:
+        return
+    data.order_factors = [o] + [
+        o1 for o1 in data.order_factors if not o.contains(o1)]
+
+@dispatch(AddBuilder, Number)
+def append_arg_Add(data, o):
+    # 3 or NaN
+    if (o is S.NaN or data.coeff is S.ComplexInfinity and
```

Comment:
Shouldn't you be able to write dispatch functions for these classes specifically.

Reply:
For `o is S.NaN` yes, for `data.something` no because it's a field of `data`.